feat(Dialog Guidance): Implement computer avatars

geoffreykwan opened the pull request:
## Changes
Add computer avatar to Dialog Guidance authoring
Add computer avatar selection to Dialog Guidance in student VLE

## Test
Authoring
- Enabling computer avatar in Dialog Guidance
- Authoring computer avatar fields (label, prompt, initial response) in Dialog Guidance
- Selecting which computer avatars to make available to students in Dialog Guidance

Student
- Make sure the student can select a computer avatar
- Make sure the computer avatar fields like label, prompt, and initial response show up properly
- Make sure choosing a computer avatar, leaving the step, and coming back, will re-populate the chosen computer avatar
- Make sure Dialog Guidance still works when computer avatar is not enabled

Grading
- Make sure computer avatars show up in Dialog Guidance grading view

geoffreykwan committed:
feat(Dialog Guidance): Implement computer avatar chooser

#491

geoffreykwan committed:
feat(Dialog Guidance): Show computer avatar in dialog responses

#491

geoffreykwan committed:
feat(Dialog Guidance): Save computer avatar in student data

#491

geoffreykwan committed:
feat(Dialog Guidance): Add authoring for computer avatar fields

- Label
- Prompt
- Initial Response

#491

geoffreykwan committed:
feat(Dialog Guidance): Show computer avatar initial response to student

#491

geoffreykwan committed:
feat(Dialog Guidance): Show computer avatar in grading view

#491

geoffreykwan committed:
feat(Dialog Guidance): Move prompt inside chat card

#491

geoffreykwan committed:
feat(Dialog Guidance): Move computer avatar images to a different folder

#491

geoffreykwan committed:
feat(Dialog Guidance): Make the computer avatar prompt bold

#491

geoffreykwan committed:
feat(Dialog Guidance): Create computer avatar service and class

#491

geoffreykwan committed:
feat(Dialog Guidance): Allow author to choose computer avatars

#491

geoffreykwan committed:
fix(Dialog Guidance): Use computer avatar in dialog response waiting msg

#491

geoffreykwan committed:
test(Dialog Guidance): Add tests for computer avatar authoring

#491

geoffreykwan committed:
test(Dialog Guidance): Add tests for computer avatar student view

#491

geoffreykwan committed:
fix(Dialog Guidance): Properly use default computer avatar values

#491

breity committed:
fix(Dialog Guidance Authoring): Save each time enable avatar checkbox is toggled
#491

breity committed:
style(Dialog Gudiance): Update style and layout to match designs

- Add student and avatar names to dialog responses
- Support html content for student avatar selector prompt

#491

breity committed:
chore(i18n): Update translation messages

- Fix some html attribute translations

#491

breity committed:
style(Dialog Guidance Authoring): Update styles, interface, accessibility

- Switch to filled mat-input fields and add hints
- Make avatar selection keyboard accessible

 #491

breity committed:
feat(Dialog Guidance): Auto-select computer avatar if only 1 option

#491

breity committed:
chore(Cleanup): Remove commented code

#491

breity committed:
Merge branch 'develop' into issue-491-dialog-guidance-avatar

breity committed:
refactor(Dialog Guidance Authoring): Specify type for array of computer avatars

#491

breity reviewed (approved):
Looks great.

I updated the styles and UI and also changed the student view so that the avatar is automatically selected if there's only 1 available (as specified in the authoring).

breity committed:
test(Computer Avatar): Fix test

- Mark default avatar label for translation

#491

hirokiterashima committed:
refactor(DG Avatar): Pass in ComputerAvatar class

- Pass in the ComputerAvatar class itself instead of id's

hirokiterashima committed:
refactor(DG Avatar): Add ComputerAvatarSettings,
EditDialogGuidanceComputerAvatar

- ComputerAvatarSettings stores configurations for ComputerAvatar such
  as prompt, label, and initialResponse.
- Extract EditDialogGuidanceComputerAvatar from
  DialogGuidanceAuthoringComponent

hirokiterashima commented:
The functionality looked good. I refactored some code so that it would be simpler and easier to read (in my opinion). I tried to pass in the ComputerAvatar object (instead of computerAvatarId) where possible, introduced ComputerAvatarSettings class, and extracted EditDialogGuidanceComputerAvatarComponent from DialogGuidanceAuthoringComponent. The component content should now look like this:
```
...
    "isComputerAvatarEnabled": true,
    "computerAvatarSettings": {
         "label": "Best friend",
        "prompt": "Discuss your answer with with your best friend!",
        "initialResponse": "Aloha! What do you think about...",
        "ids": {
            "robot": true,
            "girl": true,
            "tiger": true
        }
   }
...
```

Let me know if you have any thoughts about these changes.

@geoffreykwan 

1. When no initial avatar is selected in the authoring tool, we still save it. Should we not allow this, make sure the author chooses initial avatars?

2. Is there a reason why the component.computerAvatarIds (now stored in component.computerAvatarSettings.ids after refactoring) is an object (e.g. {"robot":true, "girl":true}) instead of just an array of ids, like ["robot", "girl"]? It doesn't look like we ever set the "robot" field to false. Wouldn't it be simpler to use an array of enabled id strings?

@breity while refactoring, I touched a lot of files. I tried to be careful copying over your styles, but I might have messed up inadvertently. Can you please check that my work kept your styles in tact? Thanks!

geoffreykwan committed:
feat(Dialog Guidance): Require author to choose at least one computer avatar

- Also make avatars easier to click for authors and students

#491

geoffreykwan commented:
> 1. When no initial avatar is selected in the authoring tool, we still save it. Should we not allow this, make sure the author chooses initial avatars?

I made a change to handle this.

> 2. Is there a reason why the component.computerAvatarIds (now stored in component.computerAvatarSettings.ids after refactoring) is an object (e.g. {"robot":true, "girl":true}) instead of just an array of ids, like ["robot", "girl"]? It doesn't look like we ever set the "robot" field to false. Wouldn't it be simpler to use an array of enabled id strings?

Having the ids as keys in a map make it easier to check/uncheck avatars in the author view and also makes it easier to find which avatars to display to students in the student view.

geoffreykwan committed:
fix(Dialog Guidance): Fix test regarding unselecting all avatars

#491

geoffreykwan committed:
feat(Dialog Guidance): Handle student view when only one computer avatar

#491

breity committed:
feat(Dialog Guidance): Update set of available avatars

- Localize avatar names
 #491

hirokiterashima committed:
refactor(DG Avatar): Use array to store avatar ids (#513)

geoffreykwan committed:
fix(Dialog Guidance): Fix scenario when there is only one computer avatar

This is to handle the change of the ids data type from object to array.

#491

hirokiterashima committed:
Updated tests to work with array computerAvatarSettings.ids

geoffreykwan committed:
test(Computer Avatar): Remove a call to console.log()

#491

hirokiterashima committed:
refactor(Dialog Guidance Avatar) (#515)

* refactor(DG Avatar): Remove highlightAvatar function

Selecting and highlighting avatar is managed by Angular Material, so
avatar.isSelected field is also not necessary

* refactor(DG Avatar): Initialize component with all available avatars

Instead of checking for null id field, set it to the full list of
available avatars.

* refactor(DG Avatar): Remove numSelectedComputerAvatars

This field is no longer used. We don't allow authors to have
numSelectedComputerAvatars=0 anymore.
Also cleaned up and consolidated functions related to numSelectedComputerAvatars

* fix(Dialog Guidance): Fix a problem with not selecting an avatar

Previously if the student clicked and slightly dragged the image in the
computer avatar it didn't select the avatar even though it should have.

Co-authored-by: Geoffrey Kwan <geoffreykwan@gmail.com>

geoffreykwan committed:
fix(Dialog Guidance): Handle old DG components without computer avatar

#491
